handler/rfc8628/storage.go

```diff
@@ -0,0 +1,55 @@
+// Copyright © 2024 Ory Corp
+// SPDX-License-Identifier: Apache-2.0
+
+package rfc8628
+
+import (
+	"context"
+
+	"github.com/ory/fosite"
+	"github.com/ory/fosite/handler/oauth2"
+)
+
+// RFC8628CoreStorage is the storage needed for the DeviceAuthHandler
+type RFC8628CoreStorage interface {
+	DeviceCodeStorage
+	UserCodeStorage
+	oauth2.AccessTokenStorage
+	oauth2.RefreshTokenStorage
+}
+
+// DeviceCodeStorage handles the device_code storage
+type DeviceCodeStorage interface {
+	// CreateDeviceCodeSession stores the device request for a given device code.
+	CreateDeviceCodeSession(ctx context.Context, signature string, request fosite.Requester) (err error)
+
+	// GetDeviceCodeSession hydrates the session based on the given device code and returns the device request.
+	// If the device code has been invalidated with `InvalidateDeviceCodeSession`, this
+	// method should return the ErrInvalidatedDeviceCode error.
+	//
+	// Make sure to also return the fosite.Requester value when returning the fosite.ErrInvalidatedDeviceCode error!
+	GetDeviceCodeSession(ctx context.Context, signature string, session fosite.Session) (request fosite.Requester, err error)
+
+	// InvalidateDeviceCodeSession is called when a device code is being used. The state of the user
+	// code should be set to invalid and consecutive requests to GetDeviceCodeSession should return the
+	// ErrInvalidatedDeviceCode error.
+	InvalidateDeviceCodeSession(ctx context.Context, signature string) (err error)
+}
+
+// UserCodeStorage handles the user_code storage
+type UserCodeStorage interface {
+	// CreateUserCodeSession stores the device request for a given user code.
+	CreateUserCodeSession(ctx context.Context, signature string, request fosite.Requester) (err error)
```

Comment:
I do see a problem with this signature:

When the signing secret is rotated, the same user code will have a different signature and we could run into problems with duplicates that the database will not be able to detect. Even worse, the users will both see the same user code on their device, and therefore everything looks correct to them. In the end they might still be able to log into a different device than they expected.

Is there any way we can solve this? If not, can you provide a rationale why this is not a problem?

Reply:
That is correct, but isn't this true for all the tokens? AFAICT the same logic is used in https://github.com/ory/fosite/blob/master/handler/oauth2/storage.go#L47 and https://github.com/ory/fosite/blob/master/handler/oauth2/flow_authorize_code_token.go#L32 (even though the `GetAuthorizeCodeSession` calls the 2nd argument `code`, a `signature` is used). 
This approach is probably not needed for user codes, but the reason for implementing this way is to keep it consistent with the rest of the fosite tokens and to allow the implementer (e.g. hydra) to have a single function/logic for handling and storing them.

Reply:
Ok, we also had some more internal discussion about user-code entropy and collisions. These were our thoughts:

A high user-code entropy is needed so that it gets hard (enough) to guess a code valid at any given point in time. Because user-codes are valid only for a short period, something like 1M guesses before a hit are probably sufficient. With rate-limiting you can obviously make it harder, but it will always be possible for someone with enough resources to circumvent the rate-limiting. For database conflicts, the same statistics apply, although a hit probability will be higher because user-codes will likely not be deleted immediately. The point here is that a production system will in a way "brute-force itself". This means that even if the chances of collisions are low, they will still happen given enough time. Here it helps to detect these rare cases early instead of presenting the user with an error later just so they have to retry.
Now considering secret rotation, it is probably OK if conflicts are not detected because it should not happen too often. However, as HMAC is relatively cheap, we would suggest to always check the codes with all or at least multiple past secrets before inserting. This could be a SQL query similar to `INSERT INTO ... SELECT 'sig1', ... WHERE NOT EXISTS (SELECT 1 FROM ... WHERE signature NOT IN ('sig1', 'sig2', 'sig3')`.
The consequence of duplicate codes is that users might be logged into different devices, but everything looks legitimate. This situation should really be avoided at any cost, and checking for duplicates during insert is a way better UX than an error when the user inputs the code and has to restart.

Reply:
We could do that also for the other code types, but I think the important part here is that the entropy of user-codes is significantly less than all other codes, and therefore we really need to consider it here.

Reply:
Thank you for the analysis, I agree with what you said. I think that the default character set we are using suffices, even in the rfc it is suggested that a smaller set can be used.

I am not sure any of this applies to the changes in fosite, as it does not implement the databases queries. Perhaps talking about this in the hydra PR would make more sense. Should I write this somewhere as guidelines for how to use the library?

Reply:
Sure, this was more about how it should be used. I think it makes sense to add some of these considerations as docs to fosite.
It does impact fosite in the way that if one would want to use the multi-signature optimization, fosite would need to support that.

Reply:
Could you point me to which doc files I should add this to?

I think that the current API allows for an implementer to check for multiple signatures (the implementation can be anything they want), unless what you want is for `UserCodeSignature` and `GenerateUserCode` to return an array of signatures and `CreateUserCodeSession`, `GetUserCodeSession` and `InvalidateUserCodeSession` to get an array of signatures as input. Can you be more specific on what you would like to change?